src/CodeViewer/Viewer.cpp

```diff
@@ -28,27 +29,37 @@ static const QColor kTextEditBackgroundColor{30, 30, 30};
 static const QColor kTextEditForegroundColor{189, 189, 189};
 static const QColor kHeatmapColor{Qt::red};
 
+int StringWidthInPixels(const QFontMetrics& font_metrics, std::string string) {
+  return font_metrics.horizontalAdvance(QString::fromStdString(string));
+}
+
 int DetermineLineNumberWidthInPixels(const QFontMetrics& font_metrics, int max_line_number) {
-  return font_metrics.horizontalAdvance(QString::number(max_line_number));
+  return StringWidthInPixels(font_metrics, std::to_string(max_line_number));
 }
 
-Viewer::Viewer(QWidget* parent) : QPlainTextEdit(parent), left_sidebar_widget_{this} {
-  UpdateLeftSidebarWidth();
-  QObject::connect(this, &QPlainTextEdit::blockCountChanged, this, &Viewer::UpdateLeftSidebarWidth);
+Viewer::Viewer(QWidget* parent)
+    : QPlainTextEdit(parent), left_sidebar_widget_{this}, right_sidebar_widget_{this} {
+  UpdateSidebarsWidth();
+  QObject::connect(this, &QPlainTextEdit::blockCountChanged, this, &Viewer::UpdateSidebarsWidth);
 
   QObject::connect(&left_sidebar_widget_, &PlaceHolderWidget::PaintEventTriggered, this,
                    &Viewer::DrawLineNumbers);
 
+  QObject::connect(&right_sidebar_widget_, &PlaceHolderWidget::PaintEventTriggered, this,
+                   &Viewer::DrawSampleCounters);
+
   const auto update_viewport_area = [&](const QRect& rect, int dy) {
     const auto update_caused_by_scroll = [](int dy) { return dy != 0; };
 
     if (update_caused_by_scroll(dy)) {
       left_sidebar_widget_.scroll(0, dy);
+      right_sidebar_widget_.scroll(0, dy);
     } else {
       QRect line_number_rect = rect;
 
       // We keep the (vertical) y-coordinates and adjust the horizontal x-coordinates for the line
       // number area.
+      // TODO: Is this really needed? Which update can break it?
```

Comment:
You are going to address this TODO within this PR, right?

Reply:
Yes

Reply:
Actually, I want @beckerhe opinion here.

Reply:
I think this is needed. At least it doesn't hurt.

Reply:
I had a closer look into the qt docs. There are two situations here:

1. We need to handle scrolling - which is clear I guess
2. We need to handle content updates (changes of the source code)

The second doesn't really appear in our use-case because it's read-only. But It's still good style to handle it here. Just imagine we implement two different styles of line-break in the future. most editors allow to visually break a line when it is too long. We could have the same. And if the user toggles that every line with such a visual line break will have twice the height from before. And that's what's updated here.

Reply:
TLDR: Yes, we need that! It's our duty to handle that case, even if it is unlikely right now, that this code path is taken.

Reply:
I changed. WDYT?

Reply:
``  QRect titles_bar = rect;
      QRect line_number_rect = rect;
      QRect samples_info_rect = rect;

      // We keep the (vertical) y-coordinates and adjust the horizontal x-coordinates for the each
      // bar's area
      titles_bar.setLeft(0);
      titles_bar.setWidth(top_bar_widget_.sizeHint().width());
      top_bar_widget_.update(line_number_rect);

      line_number_rect.setLeft(0);
      line_number_rect.setWidth(left_sidebar_widget_.sizeHint().width());
      left_sidebar_widget_.update(line_number_rect);

      samples_info_rect.setLeft(0);
      samples_info_rect.setWidth(right_sidebar_widget_.sizeHint().width());
      right_sidebar_widget_.update(line_number_rect); ``

Reply:
Reverted the change:
 
     ```
 QRect titles_bar = rect;
      QRect line_number_rect = rect;
      QRect samples_info_rect = rect;

      // We keep the (vertical) y-coordinates and adjust the horizontal x-coordinates for the each
      // bar's area
      titles_bar.setLeft(0);
      titles_bar.setWidth(top_bar_widget_.sizeHint().width());
      top_bar_widget_.update(line_number_rect);

      line_number_rect.setLeft(0);
      line_number_rect.setWidth(left_sidebar_widget_.sizeHint().width());
      left_sidebar_widget_.update(line_number_rect);

      samples_info_rect.setLeft(0);
      samples_info_rect.setWidth(right_sidebar_widget_.sizeHint().width());
      right_sidebar_widget_.update(line_number_rect);
```